From f3c3ad3df40e9e3fa117e9942473789cbf3ad231 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Tue, 15 Feb 2022 16:16:52 +0100
Subject: [PATCH 0209/1695] plat-stm32mp1: tzc400: use the device tree
 configuration

Add the usage of memory regions defined to configure the
tzc400 firewall.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 338e1cfeb23cbd16b0d7eea45b33461eecbf0e1b plat-stm32mp1: tzc400: use the device tree configuration]
Change-Id: If81ad7ee3d2261accf04a403bfcb27c81b66a890
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/plat_tzc400.c | 387 +++++++++++++++++-----
 1 file changed, 309 insertions(+), 78 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/plat_tzc400.c b/core/arch/arm/plat-stm32mp1/plat_tzc400.c
index 004c8db1a..5547224b6 100644
--- a/core/arch/arm/plat-stm32mp1/plat_tzc400.c
+++ b/core/arch/arm/plat-stm32mp1/plat_tzc400.c
@@ -7,15 +7,20 @@
 #include <config.h>
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
+#include <drivers/stm32mp_dt_bindings.h>
 #include <drivers/tzc400.h>
 #include <initcall.h>
 #include <io.h>
+#include <keep.h>
 #include <kernel/dt.h>
 #include <kernel/interrupt.h>
 #include <kernel/panic.h>
+#include <kernel/spinlock.h>
+#include <kernel/tee_misc.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <platform_config.h>
+#include <stm32_util.h>
 #include <trace.h>
 #include <util.h>
 
@@ -47,10 +52,27 @@ struct tzc_device {
 	struct stm32mp_tzc_platdata pdata;
 	struct stm32mp_tzc_driver_data *ddata;
 	struct tzc_region_config *reg;
-	bool *reg_locked;
+	uint32_t nb_reg_used;
+	unsigned int lock;
 };
 
-#define filter_mask(_width) GENMASK_32(((_width) - 1U), 0U)
+#define IS_PAGE_ALIGNED(addr)		(((addr) & SMALL_PAGE_MASK) == 0)
+#define filter_mask(_width)		GENMASK_32(((_width) - 1U), 0U)
+
+/*
+ * At TZC driver initialization, there are memory regions defined in the DT
+ * with TZC configuration information. TZC is first configured for each of
+ * these regions and each is carved out from the overall memory address range
+ * controlled by TZC. This results in a series a memory regions that, by
+ * construction, are assigned to non-secure world.
+ */
+struct tzc_region_non_sec {
+	struct tzc_region_config region;
+	SLIST_ENTRY(tzc_region_non_sec) link;
+};
+
+static SLIST_HEAD(nsec_list_head, tzc_region_non_sec) nsec_region_list =
+	SLIST_HEAD_INITIALIZER(nsec_list_head);
 
 static enum itr_return tzc_it_handler(struct itr_handler *handler __unused)
 {
@@ -66,63 +88,18 @@ static enum itr_return tzc_it_handler(struct itr_handler *handler __unused)
 }
 DECLARE_KEEP_PAGER(tzc_it_handler);
 
-static int tzc_find_region(vaddr_t base, size_t size,
-			   struct tzc_region_config *region_cfg)
-{
-	uint8_t i = 1;
-
-	while (true) {
-		if (tzc_get_region_config(i, region_cfg) != TEE_SUCCESS)
-			return 0;
-
-		if (region_cfg->base <= base &&
-		    region_cfg->top >= (base + size - 1))
-			return i;
-
-		i++;
-		region_cfg++;
-	}
-}
-
-#ifdef CFG_CORE_RESERVED_SHM
-static bool tzc_region_is_non_secure(unsigned int i, uint64_t pa, size_t size)
-{
-	struct tzc_region_config region_cfg = { };
-	uint32_t ns_cpu_mask = TZC_REGION_ACCESS_RDWR(STM32MP1_TZC_A7_ID);
-	uint32_t filters_mask = TZC_FILTERS_MASK;
-	vaddr_t base = pa;
-
-	return region_cfg->sec_attr == TZC_REGION_S_NONE &&
-		(region_cfg->ns_device_access & ns_cpu_mask) == ns_cpu_mask &&
-		region_cfg->filters == filters_mask;
-}
-#endif
-
-static bool tzc_region_is_secure(struct tzc_device *tzc_dev,
-				 struct tzc_region_config *region_cfg)
+static TEE_Result tzc_region_check_overlap(struct tzc_device *tzc_dev,
+					   const struct tzc_region_config *reg)
 {
-	uint32_t filters_mask = filter_mask(tzc_dev->ddata->nb_filters);
+	unsigned int i = 0;
 
-	return region_cfg->sec_attr == TZC_REGION_S_RDWR &&
-		region_cfg->ns_device_access == 0 &&
-		region_cfg->filters == filters_mask;
-}
-
-static void stm32mp_tzc_check_boot_region(struct tzc_device *tzc_dev)
-{
-	int idx = 0;
+	/* Check if base address already defined in another region */
+	for (i = 0; i < tzc_dev->nb_reg_used; i++)
+		if (reg->base <= tzc_dev->reg[i].top &&
+		    reg->top >= tzc_dev->reg[i].base)
+			return TEE_ERROR_ACCESS_CONFLICT;
 
-	idx = tzc_find_region(CFG_TZDRAM_START, CFG_TZDRAM_SIZE,
-			      &tzc_dev->reg[1]);
-	if (!idx || !tzc_region_is_secure(tzc_dev, &tzc_dev->reg[idx]))
-		panic("TZC: bad permissions on TEE RAM");
-
-#ifdef CFG_CORE_RESERVED_SHM
-	idx = tzc_find_region(CFG_SHMEM_START, CFG_SHMEM_SIZE,
-			      &tzc_dev->reg[1]);
-	if (!idx || !tzc_region_is_non_secure(tzc_dev, &tzc_dev->reg[idx]))
-		panic("TZC: bad permissions on TEE reserved SHMEM");
-#endif
+	return TEE_SUCCESS;
 }
 
 static void tzc_set_driverdata(struct tzc_device *tzc_dev)
@@ -172,6 +149,247 @@ static void tzc_free(struct tzc_device *tzc_dev)
 	}
 }
 
+static void stm32mp_tzc_region0(bool enable)
+{
+	struct tzc_region_config region_cfg_0 = {
+		.base = 0,
+		.top = UINT_MAX,
+		.sec_attr = TZC_REGION_S_NONE,
+		.ns_device_access = 0,
+	};
+
+	if (enable)
+		region_cfg_0.sec_attr = TZC_REGION_S_RDWR;
+
+	tzc_configure_region(0, &region_cfg_0);
+}
+
+static void stm32mp_tzc_reset_region(struct tzc_device *tzc_dev)
+{
+	unsigned int i = 0;
+	const struct tzc_region_config cfg = { .top = 0x00000FFF };
+
+	/* Clean old configuration */
+	for (i = 0; i < tzc_dev->ddata->nb_regions; i++)
+		tzc_configure_region(i + 1, &cfg);
+}
+
+static TEE_Result stm32mp_tzc_region_append(struct tzc_device *tzc_dev,
+					    const struct tzc_region_config
+					    *region_cfg)
+{
+	TEE_Result res = TEE_SUCCESS;
+	unsigned int index = 0;
+	uint32_t exceptions = 0;
+
+	exceptions = may_spin_lock(&tzc_dev->lock);
+	index = tzc_dev->nb_reg_used;
+
+	if (index >= tzc_dev->ddata->nb_regions ||
+	    region_cfg->base < tzc_dev->pdata.mem_base ||
+	    region_cfg->top > tzc_dev->pdata.mem_base +
+	    (tzc_dev->pdata.mem_size - 1)) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto end;
+	}
+
+	res = tzc_region_check_overlap(tzc_dev, region_cfg);
+	if (res)
+		goto end;
+
+	memcpy(&tzc_dev->reg[index], region_cfg,
+	       sizeof(struct tzc_region_config));
+
+	tzc_configure_region(index + 1, region_cfg);
+
+	tzc_dev->nb_reg_used++;
+
+end:
+	may_spin_unlock(&tzc_dev->lock, exceptions);
+
+	return res;
+}
+
+static TEE_Result
+exclude_region_from_nsec(const struct tzc_region_config *reg_exclude)
+{
+	struct tzc_region_non_sec *reg = NULL;
+	bool found = false;
+
+	SLIST_FOREACH(reg, &nsec_region_list, link) {
+		found = core_is_buffer_inside(reg_exclude->base,
+					      reg_exclude->top -
+					      reg_exclude->base + 1,
+					      reg->region.base,
+					      reg->region.top -
+					      reg->region.base + 1);
+		if (found)
+			break;
+	}
+
+	if (!found || !reg)
+		panic();
+
+	if (reg_exclude->base == reg->region.base &&
+	    reg_exclude->top == reg->region.top) {
+		/* Remove this entry */
+		SLIST_REMOVE(&nsec_region_list, reg, tzc_region_non_sec, link);
+	} else if (reg_exclude->base == reg->region.base) {
+		reg->region.base = reg_exclude->top + 1;
+	} else if (reg_exclude->top == reg->region.top) {
+		reg->region.top = reg_exclude->base - 1;
+	} else {
+		struct tzc_region_non_sec *new_nsec =
+			calloc(1, sizeof(*new_nsec));
+
+		if (!new_nsec || !reg)
+			panic();
+
+		memcpy((void *)&new_nsec->region, (void *)&reg->region,
+		       sizeof(struct tzc_region_config));
+		reg->region.top = reg_exclude->base - 1;
+		new_nsec->region.base = reg_exclude->top + 1;
+		SLIST_INSERT_AFTER(reg, new_nsec, link);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void stm32mp_tzc_cfg_boot_region(struct tzc_device *tzc_dev)
+{
+	unsigned int idx = 0;
+	static struct tzc_region_config boot_region[] = {
+		{
+			.base = CFG_TZDRAM_START,
+			.top = CFG_TZDRAM_START + CFG_TZDRAM_SIZE - 1,
+			.sec_attr = TZC_REGION_S_RDWR,
+			.ns_device_access = 0,
+		},
+#ifdef CFG_CORE_RESERVED_SHM
+		{
+			.base = CFG_SHMEM_START,
+			.top = CFG_SHMEM_START + CFG_SHMEM_SIZE - 1,
+			.sec_attr = TZC_REGION_S_NONE,
+			.ns_device_access =
+				TZC_REGION_ACCESS_RDWR(STM32MP1_TZC_A7_ID),
+		}
+#endif
+	};
+
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_TZDRAM_START));
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_TZDRAM_SIZE));
+#ifdef CFG_CORE_RESERVED_SHM
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_SHMEM_START));
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_SHMEM_SIZE));
+#endif
+
+	stm32mp_tzc_region0(true);
+
+	stm32mp_tzc_reset_region(tzc_dev);
+
+	for (idx = 0; idx < ARRAY_SIZE(boot_region); idx++) {
+		TEE_Result res = TEE_ERROR_GENERIC;
+
+		boot_region[idx].filters =
+			filter_mask(tzc_dev->ddata->nb_filters);
+
+		res = stm32mp_tzc_region_append(tzc_dev, &boot_region[idx]);
+		if (res)
+			panic("Enable to configure core regions");
+
+		res = exclude_region_from_nsec(&boot_region[idx]);
+	}
+
+	/* Remove region0 access */
+	stm32mp_tzc_region0(false);
+}
+
+static TEE_Result add_node_memory_regions(struct tzc_device *tzc_dev,
+					  const void *fdt, int node)
+{
+	const fdt32_t *conf_list = NULL;
+	unsigned int nregions = 0;
+	unsigned int i = 0;
+	int len = 0;
+
+	conf_list = fdt_getprop(fdt, node, "memory-region", &len);
+	if (!conf_list)
+		return TEE_SUCCESS;
+
+	nregions = len / sizeof(uint32_t);
+	if (nregions > tzc_dev->ddata->nb_regions)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < nregions; i++) {
+		struct tzc_region_config region_cfg = { };
+		int pnode = 0;
+		size_t region_size = 0;
+		paddr_t region_base = 0;
+		const fdt32_t *prop = NULL;
+		uint32_t phandle = fdt32_to_cpu(*(conf_list + i));
+
+		pnode = fdt_node_offset_by_phandle(fdt, phandle);
+		if (pnode < 0)
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		region_base = fdt_reg_base_address(fdt, pnode);
+		region_size = fdt_reg_size(fdt, pnode);
+		assert(region_base != (paddr_t)-1 && region_size != (size_t)-1);
+
+		if (!IS_PAGE_ALIGNED(region_base) ||
+		    !IS_PAGE_ALIGNED(region_size))
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		region_cfg.base = region_base;
+		region_cfg.top = region_base + region_size - 1;
+		region_cfg.filters = filter_mask(tzc_dev->ddata->nb_filters);
+
+		prop = fdt_getprop(fdt, pnode, "st,protreg", &len);
+		if (!prop || len != (2 * sizeof(uint32_t)))
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		region_cfg.sec_attr = fdt32_to_cpu(prop[0]);
+		region_cfg.ns_device_access = fdt32_to_cpu(prop[1]);
+
+		DMSG("%#08"PRIxVA" - %#08"PRIxVA" : Sec access %i NS access %#"PRIx32,
+		     region_cfg.base, region_cfg.top, region_cfg.sec_attr,
+		     region_cfg.ns_device_access);
+
+		if (stm32mp_tzc_region_append(tzc_dev, &region_cfg))
+			panic("Error adding region");
+
+		if (exclude_region_from_nsec(&region_cfg))
+			panic("Not able to exclude region");
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Adds a TZC region entry for each non-secure memory area defined by
+ * nsec_region_list. The function releases resources used to build this
+ * non-secure region list.
+ */
+static void add_carved_out_nsec(struct tzc_device *tzc_dev)
+{
+	struct tzc_region_non_sec *region = NULL;
+	struct tzc_region_non_sec *region_safe = NULL;
+
+	SLIST_FOREACH_SAFE(region, &nsec_region_list, link, region_safe) {
+		DMSG("%#08"PRIxVA" - %#08"PRIxVA" : Sec access %i NS access %#"PRIx32,
+		     region->region.base, region->region.top,
+		     region->region.sec_attr,
+		     region->region.ns_device_access);
+
+		if (stm32mp_tzc_region_append(tzc_dev, &region->region))
+			panic("Error adding region");
+
+		SLIST_REMOVE(&nsec_region_list, region,
+			     tzc_region_non_sec, link);
+		free(region);
+	};
+}
+
 static TEE_Result stm32mp_tzc_parse_fdt(struct tzc_device *tzc_dev,
 					const void *fdt, int node)
 {
@@ -213,8 +431,8 @@ static TEE_Result stm32mp_tzc_parse_fdt(struct tzc_device *tzc_dev,
 	tzc_dev->pdata.mem_base = fdt_reg_base_address(fdt, offs);
 	tzc_dev->pdata.mem_size = fdt_reg_size(fdt, offs);
 
-	assert(tzc_dev->pdata.mem_base != (paddr_t)-1 &&
-	       tzc_dev->pdata.mem_size != (size_t)-1);
+	assert(tzc_dev->pdata.mem_base != DT_INFO_INVALID_REG &&
+	       tzc_dev->pdata.mem_size != DT_INFO_INVALID_REG_SIZE);
 
 	return TEE_SUCCESS;
 }
@@ -224,35 +442,54 @@ static TEE_Result stm32mp1_tzc_probe(const void *fdt, int node,
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
 	struct tzc_device *tzc_dev = NULL;
-
-	assert(fdt && node >= 0);
+	struct tzc_region_non_sec *nsec_region = NULL;
 
 	tzc_dev = tzc_alloc();
 	if (!tzc_dev)
-		return TEE_ERROR_OUT_OF_MEMORY;
+		panic();
 
 	res = stm32mp_tzc_parse_fdt(tzc_dev, fdt, node);
-	if (res)
-		goto err;
+	if (res) {
+		tzc_free(tzc_dev);
+		return res;
+	}
 
 	tzc_set_driverdata(tzc_dev);
 	tzc_dev->reg = calloc(tzc_dev->ddata->nb_regions,
-				sizeof(*tzc_dev->reg));
-	tzc_dev->reg_locked = calloc(tzc_dev->ddata->nb_regions,
-				     sizeof(*tzc_dev->reg_locked));
-	if (!tzc_dev->reg || !tzc_dev->reg_locked) {
-		res = TEE_ERROR_OUT_OF_MEMORY;
-		goto err;
-	}
+			      sizeof(*tzc_dev->reg));
+	if (!tzc_dev->reg)
+		panic();
 
 	clk_enable(tzc_dev->pdata.clk[0]);
 	if (tzc_dev->pdata.clk[1])
 		clk_enable(tzc_dev->pdata.clk[1]);
 
 	tzc_init((vaddr_t)tzc_dev->pdata.base);
-	tzc_dump_state();
 
-	stm32mp_tzc_check_boot_region(tzc_dev);
+	nsec_region = calloc(1, sizeof(*nsec_region));
+	if (!nsec_region)
+		panic();
+
+	nsec_region->region.base = tzc_dev->pdata.mem_base;
+	nsec_region->region.top = tzc_dev->pdata.mem_base +
+				  tzc_dev->pdata.mem_size - 1;
+	nsec_region->region.sec_attr = TZC_REGION_S_NONE;
+	nsec_region->region.ns_device_access = TZC_REGION_NSEC_ALL_ACCESS_RDWR;
+	nsec_region->region.filters = filter_mask(tzc_dev->ddata->nb_filters);
+
+	SLIST_INSERT_HEAD(&nsec_region_list, nsec_region, link);
+
+	stm32mp_tzc_cfg_boot_region(tzc_dev);
+
+	res = add_node_memory_regions(tzc_dev, fdt, node);
+	if (res) {
+		EMSG("Can't add memory regions: %"PRIx32, res);
+		panic();
+	}
+
+	add_carved_out_nsec(tzc_dev);
+
+	tzc_dump_state();
 
 	res = interrupt_create_handler(tzc_dev->pdata.itr_chip,
 				       tzc_dev->pdata.itr_num, tzc_it_handler,
@@ -264,12 +501,6 @@ static TEE_Result stm32mp1_tzc_probe(const void *fdt, int node,
 	tzc_set_action(TZC_ACTION_INT);
 
 	return TEE_SUCCESS;
-
-err:
-	free(tzc_dev->reg);
-	tzc_free(tzc_dev);
-
-	return res;
 }
 
 static const struct dt_device_match tzc_secu_match_table[] = {
-- 
2.47.2

